aws/logs_monitoring/lambda_function.py

```diff
@@ -652,7 +675,35 @@ def awslogs_handler(event, context, metadata):
                 metadata[DD_HOST] = arn
 
     # Create and send structured logs to Datadog
-    for log in logs["logEvents"]:
+    # if multiline regex, iterate over log events and concatenate multiline logs into parent log
+    # child multiline logs must have matching timestamp and not match regex start pattern
+    log_events = []
+
+    if DD_MULTILINE_CLOUDWATCH_LOG_REGEX_PATTERNS and logs["logGroup"] in DD_MULTILINE_CLOUDWATCH_LOG_REGEX_PATTERNS:
+        multiline_log_timestamp = None
+        multiline_log_flag = False
+        log_group = logs["logGroup"]
+
+        for log in logs["logEvents"]:
+            if DD_MULTILINE_CLOUDWATCH_LOG_REGEX_PATTERNS[log_group].match(log["message"]):
+                multiline_log_timestamp = log["timestamp"]
+                multiline_log_flag = True
+                log_events.append(log)
+            elif log["timestamp"] != multiline_log_timestamp:
+                multiline_log_timestamp = None
+                multiline_log_flag = False
+                log_events.append(log)
+            elif multiline_log_flag:
+                # valid child multiline log, matching timestamp and multiline log flag has been toggled
+                log_events[-1]["message"] += log["message"]
+            else:
+                # edge case, multiple logs without starting pattern regex
```

Comment:
I feel like we can't fall into this case, can you describe me a use case please ?

Reply:
So i've been testing this with the cloudwatch logs from a couple very basic ruby/js lambdas, which can be found here: `https://gist.github.com/ericmustin/adb15cef1576baf5df7697db1f5ee7b7`

The use case here would be if multiple logs in a row did not contain the multiline start pattern. An example is aws lambdas automatically insert two logs to denote the `END` and `REPORT` of the lambda, as seen here: `https://gist.github.com/ericmustin/adb15cef1576baf5df7697db1f5ee7b7#file-example_output_node-log-L13-L14` , but there are probably a number of ways to hit this case